src/scene/scene.cpp

```diff
@@ -79,148 +82,122 @@ namespace {
 
 namespace openspace {
 
-Scene::Scene()
-    : _focus(SceneGraphNode::RootNodeName)
-{}
+Scene::Scene() {}
 
-Scene::~Scene() {
-    deinitialize();
-}
+Scene::~Scene() {}
 
-bool Scene::initialize() {
-    LDEBUG("Initializing SceneGraph");   
-    return true;
+void Scene::setRoot(std::unique_ptr<SceneGraphNode> root) {
+    if (_root) {
+        removeNode(_root.get());
+    }
+    _root = std::move(root);
+    _root->setScene(this);
+    addNode(_root.get());
 }
 
-bool Scene::deinitialize() {
-    clearSceneGraph();
-    return true;
+void Scene::setCamera(std::unique_ptr<Camera> camera) {
+    _camera = std::move(camera);
 }
 
-void Scene::update(const UpdateData& data) {
-    if (!_sceneGraphToLoad.empty()) {
-        OsEng.renderEngine().scene()->clearSceneGraph();
-        try {          
-            loadSceneInternal(_sceneGraphToLoad);
+Camera* Scene::camera() const {
+    return _camera.get();
+}
 
-            // Reset the InteractionManager to Orbital/default mode
-            // TODO: Decide if it belongs in the scene and/or how it gets reloaded
-            //OsEng.interactionHandler().setInteractionMode("Orbital");
+void Scene::addNode(SceneGraphNode* node, Scene::UpdateDependencies updateDeps) {
+    // Add the node and all its children.
+    node->traversePreOrder([this](SceneGraphNode* n) {
+        _topologicallySortedNodes.push_back(n);
+        _nodesByName[n->name()] = n;
+    });
+    
+    if (updateDeps) {
+        updateDependencies();
+    }
+}
 
-            // After loading the scene, the keyboard bindings have been set
-            const std::string KeyboardShortcutsType =
-                ConfigurationManager::KeyKeyboardShortcuts + "." +
-                ConfigurationManager::PartType;
+void Scene::removeNode(SceneGraphNode* node, Scene::UpdateDependencies updateDeps) {
+    // Remove the node and all its children.
+    node->traversePostOrder([this](SceneGraphNode* node) {
+        _topologicallySortedNodes.erase(
+            std::remove(_topologicallySortedNodes.begin(), _topologicallySortedNodes.end(), node),
+            _topologicallySortedNodes.end()
+        );
+        _nodesByName.erase(node->name());
+    });
+    
+    if (updateDeps) {
+        updateDependencies();
+    }
+}
 
-            const std::string KeyboardShortcutsFile =
-                ConfigurationManager::KeyKeyboardShortcuts + "." +
-                ConfigurationManager::PartFile;
+void Scene::updateDependencies() {
+    sortTopologically();
+}
 
+void Scene::sortTopologically() {
+    std::copy(_circularNodes.begin(), _circularNodes.end(), std::back_inserter(_topologicallySortedNodes));
```

Comment:
The `std::back_inserter` is potentially slow. A faster method is:
```
_topologicallySortedNodes.insert(
  _topologicallySortedNodes.end(), 
  make_move_iterator(_circularNodes.begin()),
  make_move_iterator(_circularNodes.end())
);
```

Your version has to call `push_back` many times, the other version does only zero or one reallocation and a single move.